blog-posts/en/angular-http-only-cookie.md

```diff
@@ -0,0 +1,423 @@
+
+```csharp
+[HttpGet]
+[AbpAuthorize]
+public async Task LogOut()
+{
+    // Other codes ...
+    
+    Response.Cookies.Delete("Abp.AuthToken");
+    Response.Cookies.Delete("Abp.AuthRefreshToken");
+}
+```
+
+### Create `ServerCookieController.cs`
+
+We need to create a new controller for setting tenant id cookie. Create a new controller named `ServerCookieController` in your `*.Web.Core` project and add the following code:
+
+```csharp
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+namespace AngularHttpOnlyCookieDemo.Web.Controllers;
+
+[Route("api/[controller]/[action]")]
+public class TenantIdCookieController : AngularHttpOnlyCookieDemoControllerBase
+{
+    [HttpPost]
+    public void SetTenantIdCookie(int tenantId)
+    {
+        if (tenantId == 0)
+        {
+            Response.Cookies.Delete("Abp.TenantId");
+        }
+        
+        var cookieOptions = new CookieOptions
+        {
+            HttpOnly = true,
+            Secure = true,
+            Expires = DateTimeOffset.UtcNow.AddYears(1)
+        };
+        
+        Response.Cookies.Append("Abp.TenantId", tenantId.ToString(), cookieOptions);
+    }
+}
+```
+
+## Step 2: Configure your `*.Web.Host` project
+
+We need to handle **HttpOnly** cookies in our `*.Web.Host` project.
+
+### Create `HttpOnlyMiddleware.cs`
+
+We need to create a new middleware for adding **HttpOnly** cookies to the request headers. Create a new class named `HttpOnlyMiddleware` in your `*.Web.Host` project and add the following code:
+
+```csharp
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+namespace AngularHttpOnlyCookieDemo.Web.Startup;
+
+public class HttpOnlyMiddleware
+{
+    private readonly RequestDelegate _next;
+    
+    public HttpOnlyMiddleware(RequestDelegate next)
+    {
+        _next = next;
+    }
+ 
+    public async Task Invoke(HttpContext context)
+    {
+        if (context.Request.Cookies.TryGetValue("Abp.AuthToken", out var token))
+        {
+            context.Request.Headers.Append("Authorization", "Bearer " + token);
+        }
+
+        await _next.Invoke(context);
+    }
+}
+
+public static class HttpOnlyMiddlewareExtension
+{
+    public static IApplicationBuilder UseHttpOnlyToken(this IApplicationBuilder builder)
+    {
+        return builder.UseMiddleware<HttpOnlyMiddleware>();
+    }
+}
+```
+
+### Configure `Startup.cs`
+
+We need to add the following code to `Startup.cs` file in your `*.Web.Host` project.
+
+```csharp
+// use before Authentication middleware
+app.UseHttpOnlyToken();
+
+app.UseAuthentication();
+```
+
+## Step 3: Configure your `Angular` project
+
+You need to remove cookies from your Angular application and a workaround for tenant switching.
+
+### Remove cookie methods
+
+We need to update `abp.js` file in your `src/assets/abp-web-resources` folder. 
+
+Find the `abp.auth.setToken` and `abp.auth.getToken` functions, then comment this functions.
```

Comment:
When completely commenting out the `abp.auth.getToken` method I get an error from the `TokenService`:
```
Unhandled Promise rejection: abp.auth.getToken is not a function ; Zone: <root> ; Task: Promise.then ; Value: TypeError: abp.auth.getToken is not a function
    at TokenService.getToken (abp-ng2-module.mjs:25:25)
...
```

Reply:
Hi @StarbuckSapien,

Could you comment like this, and try again?

![image](https://github.com/aspnetzero/documents/assets/56235083/5943fbf7-039e-446e-b86f-24a9c92b81d8)

Reply:
I don't get the error when I comment it like that.